lib/db/lowlevel.go

```diff
@@ -590,6 +627,123 @@ func (db *Lowlevel) gcBlocks() error {
 	return db.Compact()
 }
 
+func (db *Lowlevel) gcDeleted() error {
+	// The deleted file GC finds items that are deleted and older than
+	// deletedGCAge (default 180 days). If we are in sync on this item (no
+	// need entry) we remove it from the database, along with the
+	// corresponding remote entries.
+	//
+	// If the remote side doesn't GC deleted files then the entry will at
+	// some point pop up again. Ideally this should therefore be
+	// symmetrically configured.
+
+	if deletedGCAge <= 0 {
+		// <= 0 is explicitly disabled
+		return nil
+	}
+	if deletedGCAge < deletedGCMinAge {
+		// an explicit but not reasonable value, we adjust upwards a bit.
+		deletedGCAge = deletedGCMinAge
+	}
+
+	db.gcMut.Lock()
+	defer db.gcMut.Unlock()
+
+	t, err := db.newReadWriteTransaction()
+	if err != nil {
+		return err
+	}
+	defer t.Release()
+
+	cutoff := time.Now().Add(-deletedGCAge)
+
+	it, err := db.NewPrefixIterator([]byte{KeyTypeSequence})
+	if err != nil {
+		return err
+	}
+	defer it.Release()
+	for it.Next() {
+		// Load our version of the file. See if it's an old deleted one.
+
+		dkey := deviceFileKey(it.Value())
+		f, ok, err := t.getFileTrunc(dkey, true)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			continue
+		}
+		if !f.IsDeleted() || f.ModTime().After(cutoff) {
```

Comment:
I think this modtime is meaningless, as simply deleting an old file would gc it immediately? 
Or are we giving it an artificial mtime of now once we notice it's deleted?

Relying on users clock feels very fragile. Could we look back lets say up to half of the sequence numbers available or something like that?

Reply:
So, I *thought* we set the modtime to the time of deletion, but indeed we do not. That makes this whole thing a bit fraught. Even limiting to a fraction of the sequence space would do no good as we can always have the "rm -rf *" situation and then we should retain all the items even though they are all deletes. Back to the drawing board I guess. I could add a deletedTime attribute and let this kick in in six months, I could have the GC sweep keep entries for (date, max sequence) so we can look up a date six months in the past and only scan up to that sequence number ...

Reply:
I think setting an artificial mtime is fine, but then just cross your fingers and hope that the clocks on all devices are ok, as this would become what other devices would decide to gc on.